packages/hash/api/src/auth/index.ts

```diff
@@ -1,21 +1,73 @@
-import { Express } from "express";
+import { Express, Request, RequestHandler } from "express";
 import { AxiosError } from "axios";
 import { Session } from "@ory/client";
-import { kratosSdk } from "./ory-kratos";
-import { User } from "../model";
+import { GraphApi } from "@hashintel/hash-graph-client";
+import { Logger } from "@hashintel/hash-backend-utils/logger";
+import { getRequiredEnv } from "@hashintel/hash-backend-utils/environment";
+import { KratosUserIdentity, publicKratosSdk } from "./ory-kratos";
+import { UserModel } from "../model";
 
 declare global {
   namespace Express {
     interface Request {
       session: Session | undefined;
-      user: User | undefined;
+      user: UserModel | undefined;
     }
   }
 }
 
-const setupAuth = (app: Express) => {
+const KRATOS_API_KEY = getRequiredEnv("KRATOS_API_KEY");
+
+const requestHeaderContainsValidKratosApiKey = (req: Request): boolean =>
+  req.header("KRATOS_API_KEY") === KRATOS_API_KEY;
+
+const kratosAfterRegistrationHookHandler =
+  (params: {
+    graphApi: GraphApi;
+  }): RequestHandler<{}, {}, { identity: KratosUserIdentity }> =>
+  (req, res) => {
+    const { graphApi } = params;
+    void (async () => {
+      const {
+        body: { identity },
+      } = req;
+
+      if (!requestHeaderContainsValidKratosApiKey(req)) {
+        res
+          .status(401)
+          .send(
+            'Please provide the kratos API key using a "KRATOS_API_KEY" request header',
+          )
+          .end();
+      }
+
+      const { id: kratosIdentityId, traits } = identity;
+      const { emails } = traits;
+
+      await UserModel.createUser(graphApi, {
```

Comment:
In case this fails, would the webhook prevent user creation? I think you wrote about blocking persist in the comment in the kratos config file, but am unsure if the hooks provide some level of transactions as well? If they do, could we comment about that in this endpoint to make that clear? :)

Reply:
I know for sure that it prevents the registration flow from completing (i.e. the user is not logged in, and an "INTERNAL_SERVER_ERROR" is returned to the FE.

I'm unsure whether the user identity is created in the datastore, this is what I was trying to make clear in [this thread](https://github.com/hashintel/hash/pull/934#discussion_r945658052)

Reply:
I can spend more time trying to figure this out locally if we think it blocks merging this PR

Reply:
I see. I don't think it should block this PR, was just interested to know if there was some behaviour in the case of a non-ok response code

Reply:
Just to update I've tested this locally, and from what I can tell when an error happens on this line, the registration flow fails (`INTERNAL SERVER ERROR` is thrown in the browser), but the identity is persisted in the kratos pg instance. Ideally we would be able to make this web-hook trigger "before persisting", as the docs suggest you can do.

I've unfortunately not been able to find how to configure this, but I'll keep looking.

Reply:
Further update, I've found [the PR](https://github.com/ory/kratos/pull/2343) where this is being implemented, and it hasn't been merged yet :/

Reply:
I believe the identity is created when registration fails.
I ran into issues with docker not being able to resolve the hostname while registering (resulting int 500s), and when I try to re-register with the same email, I get a `An account with the same identifier (email, phone, username, ...) exists already.` error text under the sign up button

Reply:
Looks like in a fork of kratos [another team has implemented this](https://github.com/Wikia/kratos/pull/38)

Reply:
Would it be difficult to temporarily work around this by intercepting the failure and deleting the identity?

Reply:
Ahmad and I have attempted to do this in https://github.com/hashintel/hash/pull/934/commits/087343ba5e804a70224d04cd12581c9b6df89b1d.

I've also created [a follow up task](https://app.asana.com/0/0/1202795672151787/f), which reminds us to remove this temporary solution in favour of the "pre-persist" hook configuration (once it's supported by kratos).